File: leetcode/Combination_Sum_II.java
import java.util.*;

class Candidate {
    int val, num;
}

public class Solution {
    void search(Candidate[] cand, int num_cand, int target, int cur, int sum, ArrayList<Integer> in, ArrayList<ArrayList<Integer>> ret) {
        if (sum == target) {
            ret.add(new ArrayList<Integer>(in));
            return;
        }
        if (cur == num_cand) return;
        for (int i = 0; i <= cand[cur].num; ++i)
            if (sum + i * cand[cur].val <= target) {
                for (int k = 1; k <= i; ++k) in.add(cand[cur].val);
                search(cand, num_cand, target, cur + 1, sum + i * cand[cur].val, in, ret);
                for (int k = 1; k <= i; ++k) in.remove(in.size() - 1);
            }
    }
    public ArrayList<ArrayList<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        int num_cand = 0;
        Candidate[] cand = new Candidate[candidates.length];
        for (int i = 0; i < candidates.length; ++i)
            if (i == 0 || candidates[i] != candidates[i - 1]) {
                cand[num_cand] = new Candidate();
                cand[num_cand].val = candidates[i];
                cand[num_cand].num = 1;
                ++num_cand;
            } else {
                ++cand[num_cand - 1].num;
            }
        ArrayList<Integer> in = new ArrayList<Integer>();
        ArrayList<ArrayList<Integer>> ret = new ArrayList<ArrayList<Integer>>();
        search(cand, num_cand, target, 0, 0, in, ret);
        return ret;
    }
}

File: leetcode/Combination_Sum_II.cpp
class Solution {
    vector<vector<int> > combinations;
    void generate(int k, vector<int> &candidates, int sum, 
                  vector<int> &combination, int target) {
        if (sum == target) {
            combinations.push_back(combination);
            return;
        }
        if (k == candidates.size())
            return;
        if (sum + candidates[k] > target)
            return;
        // Use k-th element
        combination.push_back(candidates[k]);
        generate(k + 1, candidates, sum + candidates[k], combination, target);
        combination.pop_back();
        // Not use k-th element
        int next_k = k + 1;  // In order to avoid duplicate combination
        while (next_k < candidates.size() && 
               candidates[next_k] == candidates[k])
            ++next_k;
        generate(next_k, candidates, sum, combination, target);
    }
public:
    vector<vector<int> > combinationSum2(vector<int> &num, int target) {
        combinations.clear();
        sort(num.begin(), num.end());
        vector<int> combination;
        generate(0, num, 0, combination, target);
        return combinations;
    }
};
